src/card.rs

```diff
@@ -0,0 +1,260 @@
+
+#[cfg(feature = "serialize")]
+use arrayvec::ArrayVec;
+#[cfg(feature = "serialize")]
+use bincode;
+#[cfg(feature = "serialize")]
+use serde::{Deserialize, Serialize};
+use core::i32;
+use core::result;
+
+const MIN_SAVE_AMNT: usize = 0x2000;
+const FILE_NAME_MAX: usize = 32;
+const CARD_ICON_MAX: usize = 8;
+#[allow(dead_code)]
+const CARD_ICON_WIDTH: usize = 32;
+#[allow(dead_code)]
+const CARD_ICON_HEIGHT: usize = 32;
+#[allow(dead_code)]
+const CARD_BANNER_WIDTH: usize = 96;
+#[allow(dead_code)]
+const CARD_BANNER_HEIGHT: usize = 32;
+
+macro_rules! card_try {
+    ($e:expr, $id:expr) => {
+        let result: CardError = unsafe { $e.into() };
+        if result == CardError::Ready {
+            report!("Card {}: {:#?}", $id, result);
+        } else {
+            report!("Card {} failed. -> {:?}", $id, result);
+            return Err(result);
+        }
+    };
+}
+
+#[derive(PartialEq, Debug)]
+pub enum CardError {
+    Unlocked = 1,
+    Ready = 0,
+    Busy = -1,
+    Wrongdevice = -2,
+    Nocard = -3,
+    Nofile = -4,
+    IOerror = -5,
+    Broken = -6,
+    Exist = -7,
+    Noent = -8,
+    Insspace = -9,
+    Noperm = -10,
+    Limit = -11,
+    Nametoolong = -12,
+    Encoding = -13,
+    Canceled = -14,
+    FatalError = -128,
+}
+
+impl From<i32> for CardError {
+    fn from(i: i32) -> CardError {
+        match i {
+            1 => CardError::Unlocked,
+            0 => CardError::Ready,
+            -1 => CardError::Busy,
+            -2 => CardError::Wrongdevice,
+            -3 => CardError::Nocard,
+            -4 => CardError::Nofile,
+            -5 => CardError::IOerror,
+            -6 => CardError::Broken,
+            -7 => CardError::Exist,
+            -8 => CardError::Noent,
+            -9 => CardError::Insspace,
+            -10 => CardError::Noperm,
+            -11 => CardError::Limit,
+            -12 => CardError::Nametoolong,
+            -13 => CardError::Encoding,
+            -14 => CardError::Canceled,
+            -128 => CardError::FatalError,
+            _ => unreachable!(),
+        }
+    }
+}
+
+#[repr(C)]
+#[derive(Debug, Default)]
+pub struct CardInfo {
+    chan: i32,
+    file_num: i32,
+    offset: i32,
+    length: i32,
+    i_block: u16,
+}
+
+#[repr(C)]
+#[derive(Debug, Default)]
+pub struct CardStat {
+    // Read-only (Set by CARDGetStatus)
+    file_name: [u8; FILE_NAME_MAX],
+    length: u32,
+    game_name: [u8; 4],
+    company: [u8; 2],
+
+    // Read/Write (Set by CARDGetStatus/CardSetStatus)
+    banner_format: u8,
+    icon_addr: u32,
+    icon_format: u16,
+    icon_speed: u16,
+    commend_addr: u32,
+
+    // Read-only (Set by CARDGetStatus)
+    offset_banner: u32,
+    offset_banner_tlut: u32,
+    offset_icon: [u32; CARD_ICON_MAX],
+    offset_icon_tlut: u32,
+    offset_data: u32,
+}
+
+extern "C" {
+    #[link_name = "CARDOpen"]
+    pub fn card_open(chan: i32, name: *const u8, card_info: *mut CardInfo) -> i32;
```

Comment:
Considering this is in a card module, it would probably be preferable to just name these without the prefix, so one can either do card::open or just import the functions directly.